### DISCUSSION TITLE
Direct canonization of a SMILES code

### DISCUSSION THREAD

**Discussion by fjvico** [rdkit/rdkit#6405] [Q&A]
A SMILES code can be canonized by obtaining the represented compound:
```python
molecule = Chem.MolFromSmiles(SMILEScode)
canonicalSMILEScode = Chem.rdmolfiles.MolToSmiles(molecule)
```
Is there any way to convert a non-canonical string to a canonical string without obtaining the molecule? The motivation is accelerating computation.

**Comment by thegodone** [comment]
This cannot be done cause the SMILES is just a list of letters, to make a
canonization you need to have the molecular graph to determine the output
SMILES after that the procedure of "canonisation" applied on molecular
graph.

**Reply by fjvico** [reply]
May be for non-cyclic (or non-aromatic) compounds, this can be done at the string level. Don't know if this would result in a faster algorithm. In essence, the information is always there (since the molecule can be determined form the SMILES string), only that a graph is probably a more convenient object to perform the maths than a linear representation.

**Reply by bp-kelley** [reply]
I'm not sure that's entirely true, you can define rather complicated trees that still need to be parsed into a graph.

```
C(CCC(CC)(CC(CCC(CC)(CC)C)))
```

**Comment by bp-kelley** [comment]
<del>You might look at CanonSmiles</del> 

completely misunderstood the question :)

```
>>> from rdkit import Chem
>>> Chem.CanonSmiles("CCN")
'CCN'
```

This raises a Boost.Python.ArgumentError on failure, so you should wrap it
in a try...except

**Reply by bp-kelley** [reply]
This does, however, create a molecule internally.

**Comment by thegodone** [comment]
Bryan, is this not just a python wrapper to mol object ?

Le mar. 30 mai 2023 à 15:56, Brian Kelley ***@***.***> a
écrit :

> You might look at CanonSmiles
>
> ```
> >>> from rdkit import Chem
> >>> Chem.CanonSmiles("CCN")
> 'CCN'
> ```
>
> This raises a Boost.Python.ArgumentError on failure, so you should wrap it
> in a try...except
>
>

**Reply by bp-kelley** [reply]
It sure does, I misunderstood the question.

**Comment by thegodone** [comment]
In fact in my papers in NLP https://link.springer.com/chapter/10.1007/978-3-030-30493-5_79, https://arxiv.org/abs/2003.02804 : we did a transformer to make the conversion of random SMILES to canonical SMILES and it is not a simple task. I would be surprised
If we can do it without using the molecular graph. I know that is transformer solution is slower than using rdkit anyway, maybe the recent work on ChatGPT can improve the speed of inference of our transformer, but transformer embedding is quite interesting for learning extra tasks la classification or regression.
